### PR TITLE
Add multipart/form-data response builders to axum-extra

### LINKED ISSUE #2624: Add a native way to create and respond to requests with multipart forms
- [x] I have looked for existing issues (including closed) about this

### Motivation

Currently, as far as I can tell, Axum does not provide a native way to respond to requests with multipart forms. 

An endpoint might return a multipart form for a variety of reasons, but the most common being the need to include multiple pieces of data (a json, a file, plaintext) in a single request.

For example, I would like to send a json and a file in a single response.

### Proposal
I am not comfortable with the Axum codebase, so I am unsure of the most idiomatic ways do this, but there are a few options that may work:

Implement `IntoResponse` for the `Multipart` extractor, and add methods like `add_field()` to the existing impls for multipart.

It may also make more sense to create a separate MultipartBuilder and implement `IntoResponse` for that type, and add builder methods, [Reqwest's implementation](https://github.com/tokio-rs/axum/blob/main/examples/reqwest-response/src/main.rs) could be used as a reference.

I don't think there are any significant drawbacks, it will marginally increase the complexity of exposed formdata APIs, but I believe the benefits outweigh the drawbacks.

### Alternatives
As far as I can tell, alternatives involve:
- Having users wrap `reqwest`'s implementation and manually interop
  - This adds a level of complexity, for no real advantages from an interface perspective.
- Having users entirely create multipart form builders from scratch.

I am willing to attempt to implement this, however I am not confident in my abilities to do so.

### PR DESCRIPTION
**Pull request by zleyyij**
Closes #2624 (mine)
This pull request provides builders that enable `multipart/form-data` forms to be returned from handlers. 

## Motivation
`multipart/form-data` forms provide a way to include multiple types of data in a single response as key value pairs. You could include an image and a JSON, or any other combination of data. 

As an example, this could be used for a pastebin, so that a single request to an axum backend could return metadata about a paste in a json, and the paste itself. The paste wouldn't need to conform to UTF-8, and could be sent separately from the json.

## Solution
I added a `MultipartForm` struct that implements `IntoResponse`, and implementers can add parts to that form by defining `Part`s. I did not implement `IntoResponse` on the `Multipart` extractor, as advised in the original issue.

One dependency was added if the `multipart` feature is enabled. `fastrand` is used to generate boundaries, mirroring Reqwest's implementation.

Thank you for your time :)

### CODE REVIEW

**Comment by zleyyij** [comment]
I may need some help fixing those CI errors, I am unsure how to address them

**Comment by zleyyij** [comment]
Ok, I managed to fix all of those CI errors, thank you for your patience

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:17]:
```suggestion
/// This struct implements [IntoResponse], and so it can be returned from a handler.
```

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:56]: I think this should be named so it's clearer what it does, e.g. `add_part` or `push_part`.

Personally, I'm not also sure this API is that useful,wouldn't it be better to just have something like `parts_mut` which would return a mutable reference to the parts vector and then the user can just push there directly, but also remove stuff, clear it, or any other operation they might need to take.

On the other hand they can just build the vector beforehand and then call `from_parts` just before returning the response.

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:136]: Generally, it makes sense to require `String` if that's what you'll need.

Currently, if someone had a large string content, instead of moving it into this constructor (assuming the caller doesn't need it anymore), they would pass only slice here and this method would clone it for no reason.

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:140]: We already depend on `mime` so I'd suggest using that to construct the string. It could avoid some typo bugs. It would be great if it would be possible to use it also for the multi part content type. I don't know if it's possible with the dynamic boundary though.

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:11]: Defaults by us choosing it? You're setting it explicitly in the `Part::text` for example.

So I think this should be named `TextUtf8` or something se more descriptive.

If you want to signal somewhere you're using just a default that might change one day, derive `Default` and use that.

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:264]: Is this also copied from reqwest?

So we take system time, do 3 xors and shifts and repeat. I would be very surprised if this took a millisecond, I would even expect all of the four "pseudorandom" numbers to be the same one most of the time. Then again it's not really that probable the data would contain the same random `u64` in hex four times next to each other separated by a hyphen. So it serves its purpose.

If you want to keep this, just use the last result as the next seed I guess. But I don't think that using `rand` (or `fastrand`) would hurt performance here much. This seems like a case of premature optimization.

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:278]: You don't need to write why you import something. Personally, I would recommend against it because it's just a matter of time before someone needs to change something here and suddenly the comments are wrong.

File: axum-extra/src/lib.rs
Comment (mladedav) [inline axum-extra/src/lib.rs:87]: Why "builder"?

File: axum-extra/src/multipart_builder.rs
Comment (mladedav) [inline axum-extra/src/multipart_builder.rs:71]: I would caution against unwraps. I think this should not happen, but either write a comment about why it's totally ok or just bite the bullet and return an explicit 500 here on error.

**Review by mladedav — DISMISSED** [review]
Thanks, for the PR I think the functionality looks good. I have not personally worked with multipart on this level so I don't know much of the technical details but I did comment on some of the other stuff.

File: axum-core/Cargo.toml
Comment (zleyyij) [reply axum-core/Cargo.toml:37]: I have no clue, I don't remember ever adding that.

File: axum-extra/src/multipart_builder.rs
Comment (zleyyij) [reply axum-extra/src/multipart_builder.rs:56]: You're right, it did feel kind of odd, I was trying to create something similar to  Reqwest's implementation (`.part()`) chained calls, but with `.with_parts()` it feels redundant, so I removed it. I can add `parts_mut` if you feel it would prudent.

File: axum-extra/src/multipart_builder.rs
Comment (zleyyij) [reply axum-extra/src/multipart_builder.rs:264]: No, they use their own implementation of an xor shift
https://github.com/seanmonstar/reqwest/blob/7e7011a657e2f97d3b52dda7bcd8af4353c06b8a/src/util.rs#L27

This impl is definitely a bit jank.

I used this here because frankly I tried to use `fastrand` first, but couldn't figure out how to make it work without CI errors. I am not experienced at all working with `cargo.toml` at this scale. I'll dig into that some more and try to figure it out, thanks for the feedback

File: axum-extra/src/lib.rs
Comment (zleyyij) [reply axum-extra/src/lib.rs:87]: There's already a `Multipart` extractor, so I wanted to clarify between that and the functionality for creating forms.

File: axum-extra/src/multipart_builder.rs
Comment (zleyyij) [reply axum-extra/src/multipart_builder.rs]: I figured this out, the impl uses fastrand now.

File: axum-extra/src/multipart_builder.rs
Comment (zleyyij) [reply axum-extra/src/multipart_builder.rs]: Right now all of the `Part::*` constructors return Self directly, but `Part::raw_part` relies on a mime type passed in as an argument. 

I know that it's generally considered bad practice to add third party crates to the public api without good cause, but the only other option I can see here is having the function internally parse the provided string into a valid mime type, which should return a result, making this portion of the interface a little bit clunkier.

Thoughts? I could include mime in the public interface, or change the interface to return a result, making it less ergonomic.

**Comment by zleyyij** [comment]
I have commit ready to fix some of those minor issues, I just need to figure out how to handle mime type stuff from a public api level

**Comment by zleyyij** [comment]
I don't know how to do that, any advice would be appreciated

**Comment by zleyyij** [comment]
I gave up and just made any mime parsing happen internally, and the associated functions return an error.

I ran as much CI as I could with `act` locally, hopefully it passes

**Comment by zleyyij** [comment]
Is there anything more this PR needs?

**Comment by jplatte** [comment]
No, it's just waiting on my review. I've started looking at it now.

File: axum-extra/src/multipart_builder.rs
Comment (jplatte) [inline axum-extra/src/multipart_builder.rs:28]: What use is this if there's no method to add a part?

File: axum-extra/src/multipart_builder.rs
Comment (jplatte) [inline axum-extra/src/multipart_builder.rs:40]: Maybe there should be an `impl FromIterator<Part>`? Then you could do

```rust
MultipartForm::from_iter([
    first_part,
    second_part,
    third_part,
)]
```

or

```rust
some_iterator
    .map(|item| /* turn item into Part */)
    .collect()
```

File: axum-extra/src/multipart_builder.rs
Comment (jplatte) [inline axum-extra/src/multipart_builder.rs:100]: It's hard to tell where the regular comment ends and the doc comment starts. Maybe move the regular comment inside the definition and separate it from the first field's docs by an empty line?

File: axum-extra/src/multipart_builder.rs
Comment (jplatte) [inline axum-extra/src/multipart_builder.rs:227]: Are you sure that this is a thing? I googled and found `Content-Encoding` and `Transfer-Encoding` docs, but the only mention of `Content-Transfer-Encoding` was in relation to some old RFC that has been replaced since (AFAIU).

File: axum-extra/src/lib.rs
Comment (zleyyij) [reply axum-extra/src/lib.rs:86]: That's definitely an improvement IMO, I felt multipart_builder was verbose but didn't know what else to do

File: axum-extra/src/multipart_builder.rs
Comment (zleyyij) [reply axum-extra/src/multipart_builder.rs:28]: The ability to add a part was removed in a previous iteration, so this is stale, good catch :)

File: axum-extra/src/multipart_builder.rs
Comment (zleyyij) [reply axum-extra/src/multipart_builder.rs:227]: So I actually implemented most of this using RFC 2388 (whoops), which has been superceded by RFC 7578, which deprecates that field. I updated the comments and code to follow RFC 5758

**Comment by zleyyij** [comment]
Those changes have been applied, thank you for your feedback :)

File: axum-extra/src/response/multiple.rs
Comment (jplatte) [inline axum-extra/src/response/multiple.rs:207]:
```suggestion
/// A boundary is defined as a user defined (arbitrary) value that does not occur in any of the data.
///
```

File: axum-extra/src/response/multiple.rs
Comment (jplatte) [inline axum-extra/src/response/multiple.rs:141]:
```suggestion
    /// Create a new part with more fine-grained control over the semantics of that part.
    ///
    /// The caller is assumed to have set a valid MIME type.
```

File: axum-extra/src/response/multiple.rs
Comment (jplatte) [inline axum-extra/src/response/multiple.rs:117]:
```suggestion
    /// Create a new part containing a generic file, with a `Content-Type` of `application/octet-stream`
    /// using the provided file name, field name, and contents.
    ///
    /// If the MIME type of the file is known, consider using `Part::raw_part`.
```

File: axum-extra/src/response/multiple.rs
Comment (jplatte) [inline axum-extra/src/response/multiple.rs:94]:
```suggestion
    /// Create a new part with `Content-Type` of `text/plain` with the supplied name and contents.
    ///
    /// This form will not have a defined file name.
```

**Review by jplatte — APPROVED** [review]
I don't have the energy to do another review, but I think this has seen enough attention that we should just merge it by now.

There's just some doc comment rewrapping I wanna apply before merging (can likely commit it myself).

**Comment by jplatte** [comment]
Do you want this to be released in another axum-extr 0.9.x patch release? It's not a huge effort but if you don't care, I'll let it sit until I have another reason to make a patch release, or it's time for 0.10.0.

**Comment by zleyyij** [comment]
Feel free to let it sit, this is fairly low priority for me.

Thanks :)

File: axum-extra/src/response/multiple.rs
Comment (jplatte) [inline axum-extra/src/response/multiple.rs:27]: I just found this for another reason... Why did we add a new deprecated item here?? @zleyyij

Comment (zleyyij) [reply axum-extra/src/response/multiple.rs:27]: That definitely shouldn't be there, I'm so sorry. I probably added it during a refactor with the intention of changing that method before deciding it would be better if I didn't.

I'll remove it.

Thank you so much for your patience, I'm so sorry about all the hassle this PR has created

Comment (zleyyij) [reply axum-extra/src/response/multiple.rs:27]: Will be resolved by https://github.com/tokio-rs/axum/pull/2957

**Pull request merged**
